File: server/camcops_server/tasks/paradise24.py
#!/usr/bin/env python

"""
camcops_server/tasks/paradise24.py

===============================================================================

    Copyright (C) 2012, University of Cambridge, Department of Psychiatry.
    Created by Rudolf Cardinal (rnc1001@cam.ac.uk).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <https://www.gnu.org/licenses/>.

===============================================================================

**PARADISE 24 task.**

"""

from typing import Any, Dict, Optional, Type, Tuple

from cardinal_pythonlib.stringfunc import strseq
from sqlalchemy.ext.declarative import DeclarativeMeta
from sqlalchemy.sql.sqltypes import Integer

from camcops_server.cc_modules.cc_constants import CssClass
from camcops_server.cc_modules.cc_db import add_multiple_columns
from camcops_server.cc_modules.cc_html import tr_qa, tr, answer
from camcops_server.cc_modules.cc_request import CamcopsRequest
from camcops_server.cc_modules.cc_task import TaskHasPatientMixin, Task


class Paradise24Metaclass(DeclarativeMeta):
    def __init__(
        cls: Type["Paradise24"],
        name: str,
        bases: Tuple[Type, ...],
        classdict: Dict[str, Any],
    ) -> None:

        add_multiple_columns(
            cls,
            cls.Q_PREFIX,
            cls.FIRST_Q,
            cls.LAST_Q,
            coltype=Integer,
            minimum=0,
            maximum=2,
            comment_fmt="Q{n} - {s}",
            comment_strings=[
                "rested 0-2 (none - a lot)",
                "loss interest 0-2 (none - a lot)",
                "appetite 0-2 (none - a lot)",
                "sleeping 0-2 (none - a lot)",
                "irritable 0-2 (none - a lot)",
                "slowed down 0-2 (none - a lot)",
                "sad 0-2 (none - a lot)",
                "worry 0-2 (none - a lot)",
                "cope 0-2 (none - a lot)",
                "pain 0-2 (none - a lot)",
                "concentrating 0-2 (none - a lot)",
                "remembering 0-2 (none - a lot)",
                "decisions 0-2 (none - a lot)",
                "conversation 0-2 (none - a lot)",
                "walking 0-2 (none - a lot)",
                "grooming 0-2 (none - a lot)",
                "sexual 0-2 (none - a lot)",
                "staying by yourself 0-2 (none - a lot)",
                "health 0-2 (none - a lot)",
                "friendship 0-2 (none - a lot)",
                "getting along 0-2 (none - a lot)",
                "work or school 0-2 (none - a lot)",
                "money 0-2 (none - a lot)",
                "community 0-2 (none - a lot)",
            ],
        )

        super().__init__(name, bases, classdict)


class Paradise24(TaskHasPatientMixin, Task, metaclass=Paradise24Metaclass):
    __tablename__ = "paradise24"
    shortname = "PARADISE 24"

    Q_PREFIX = "q"
    FIRST_Q = 1
    LAST_Q = 24

    ALL_FIELD_NAMES = strseq(Q_PREFIX, FIRST_Q, LAST_Q)

    @staticmethod
    def longname(req: CamcopsRequest) -> str:
        _ = req.gettext
        return _(
            "Psychosocial fActors Relevant to BrAin DISorders in Europe–24"
        )

    def is_complete(self) -> bool:
        if self.any_fields_none(self.ALL_FIELD_NAMES):
            return False

        return True

    def total_score(self) -> Optional[int]:
        if not self.is_complete():
            return None

        return self.sum_fields(self.ALL_FIELD_NAMES)

    def metric_score(self) -> Optional[int]:
        total_score = self.total_score()

        if total_score is None:
            return None

        # Table 3 of Cieza et al. (2015); see help.
        # - doi:10.1371/journal.pone.0132410.t003
        # - Indexes are raw scores.
        # - Values are transformed scores.
        score_lookup = [
            0,  # 0
            10,
            19,
            25,
            29,
            33,
            36,
            38,
            41,
            43,
            45,  # 10
            46,
            48,
            50,
            51,
            53,
            54,
            55,
            57,
            58,
            59,  # 20
            60,
            61,
            63,
            64,
            65,
            66,
            67,
            68,
            69,
            71,  # 30
            72,
            73,
            74,
            76,
            77,
            78,
            80,
            81,
            83,
            85,  # 40
            87,
            89,
            91,
            92,
            94,
            96,
            98,
            100,  # 48
        ]

        try:
            return score_lookup[total_score]
        except (IndexError, TypeError):
            return None

    def get_task_html(self, req: CamcopsRequest) -> str:
        rows = ""
        for q_num in range(self.FIRST_Q, self.LAST_Q + 1):
            field = self.Q_PREFIX + str(q_num)
            question_cell = f"{q_num}. {self.xstring(req, field)}"

            rows += tr_qa(question_cell, self.get_answer_cell(req, q_num))

        html = """
            <div class="{CssClass.SUMMARY}">
                <table class="{CssClass.SUMMARY}">
                    {tr_is_complete}
                    {total_score}
                    {metric_score}
                </table>
            </div>
            <table class="{CssClass.TASKDETAIL}">
                <tr>
                    <th width="60%">Question</th>
                    <th width="40%">Score</th>
                </tr>
                {rows}
            </table>
            <div class="{CssClass.FOOTNOTES}">
                [1] Sum of all questions, range 0–48.
                [2] Transformed metric scale, range 0–100.
            </div>
        """.format(
            CssClass=CssClass,
            tr_is_complete=self.get_is_complete_tr(req),
            total_score=tr(
                self.wxstring(req, "raw_score") + " <sup>[1]</sup>",
                f"{answer(self.total_score())}",
            ),
            metric_score=tr(
                self.wxstring(req, "metric_score") + " <sup>[2]</sup>",
                f"{answer(self.metric_score())}",
            ),
            rows=rows,
        )
        return html

    def get_answer_cell(self, req: CamcopsRequest, q_num: int) -> str:
        q_field = self.Q_PREFIX + str(q_num)

        score = getattr(self, q_field)
        meaning = self.get_score_meaning(req, score)
        answer_cell = f"{score} [{meaning}]"

        return answer_cell

    def get_score_meaning(self, req: CamcopsRequest, score: int) -> str:
        return self.wxstring(req, f"option_{score}")

File: tablet_qt/tasks/paradise24.cpp
/*
    Copyright (C) 2012, University of Cambridge, Department of Psychiatry.
    Created by Rudolf Cardinal (rnc1001@cam.ac.uk).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <https://www.gnu.org/licenses/>.
*/

#include "paradise24.h"

#include "lib/convert.h"
#include "lib/stringfunc.h"
#include "maths/mathfunc.h"
#include "questionnairelib/namevaluepair.h"
#include "questionnairelib/questionnaire.h"
#include "questionnairelib/questionnairefunc.h"
#include "questionnairelib/quheading.h"
#include "questionnairelib/qumcqgrid.h"
#include "tasklib/taskfactory.h"
#include "tasklib/taskregistrar.h"
using mathfunc::anyNull;
using mathfunc::meanOrNull;
using mathfunc::sumInt;
using stringfunc::strnumlist;
using stringfunc::strseq;

const int FIRST_Q = 1;
const int LAST_Q = 24;
const int MIN_QUESTION_SCORE = 0;
const int MAX_QUESTION_SCORE = 2;
const int MIN_RAW_TOTAL_SCORE = 0;
const int MAX_RAW_TOTAL_SCORE = 48;
const int MIN_METRIC_SCORE = 0;
const int MAX_METRIC_SCORE = 100;

const QString Q_PREFIX("q");
const QString Paradise24::PARADISE24_TABLENAME("paradise24");

void initializeParadise24(TaskFactory& factory)
{
    static TaskRegistrar<Paradise24> registered(factory);
}

Paradise24::Paradise24(
    CamcopsApp& app, DatabaseManager& db, const int load_pk
) :
    Task(app, db, PARADISE24_TABLENAME, false, false, false)
// ... anon, clin, resp
{
    addFields(strseq(Q_PREFIX, FIRST_Q, LAST_Q), QMetaType::fromType<int>());

    load(load_pk);  // MUST ALWAYS CALL from derived Task constructor.
}

// ============================================================================
// Class info
// ============================================================================

QString Paradise24::shortname() const
{
    return "PARADISE 24";
}

QString Paradise24::longname() const
{
    return tr("Psychosocial fActors Relevant to BrAin DISorders in Europe–24");
}

QString Paradise24::description() const
{
    return tr(
        "A measure to assess the the impact of brain disorders on people’s "
        "lives, based on psychosocial difficulties that are experienced in "
        "common across brain disorders."
    );
}

QStringList Paradise24::fieldNames() const
{
    auto field_names = strseq(Q_PREFIX, FIRST_Q, LAST_Q);

    return field_names;
}

// ============================================================================
// Instance info
// ============================================================================


bool Paradise24::isComplete() const
{
    if (anyNull(values(fieldNames()))) {
        return false;
    }

    return true;
}

QVariant Paradise24::rawTotalScore() const
{
    if (!isComplete()) {
        return QVariant();
    }

    return sumInt(values(fieldNames()));
}

QVariant Paradise24::metricScore() const
{
    auto total_score = rawTotalScore();
    if (total_score.isNull()) {
        return QVariant();
    }

    // Table 3 of Cieza et al. (2015); see help.
    // - doi:10.1371/journal.pone.0132410.t003
    // - Indexes are raw scores.
    // - Values are transformed scores.
    const QVector<int> score_lookup = {
        0,  // 0
        10,  19, 25, 29, 33, 36, 38, 41, 43,
        45,  // 10
        46,  48, 50, 51, 53, 54, 55, 57, 58,
        59,  // 20
        60,  61, 63, 64, 65, 66, 67, 68, 69,
        71,  // 30
        72,  73, 74, 76, 77, 78, 80, 81, 83,
        85,  // 40
        87,  89, 91, 92, 94, 96, 98,
        100,  // 48
    };

    return score_lookup[total_score.toInt()];
}

QStringList Paradise24::summary() const
{
    auto rangeScore = [](const QString& description,
                         const QVariant score,
                         const int min,
                         const int max) {
        return QString("%1: <b>%2</b> [%3–%4].")
            .arg(
                description,
                convert::prettyValue(score),
                QString::number(min),
                QString::number(max)
            );
    };
    return QStringList{
        rangeScore(
            xstring("raw_score"),
            rawTotalScore(),
            MIN_RAW_TOTAL_SCORE,
            MAX_RAW_TOTAL_SCORE
        ),
        rangeScore(
            xstring("metric_score"),
            metricScore(),
            MIN_METRIC_SCORE,
            MAX_METRIC_SCORE
        ),
    };
}

QStringList Paradise24::detail() const
{
    QStringList lines = completenessInfo();

    const QString spacer = " ";
    const QString suffix = "";

    const QStringList fieldnames = fieldNames();

    for (int i = 0; i < fieldnames.length(); ++i) {
        const QString& fieldname = fieldnames.at(i);
        lines.append(
            fieldSummary(fieldname, xstring(fieldname), spacer, suffix)
        );
    }


    lines.append("");
    lines += summary();

    return lines;
}

OpenableWidget* Paradise24::editor(const bool read_only)
{
    NameValueOptions options;

    for (int i = MIN_QUESTION_SCORE; i <= MAX_QUESTION_SCORE; ++i) {
        auto name = QString("option_%1").arg(i);

        options.append({xstring(name), i});
    }

    const int min_width_px = 100;
    const QVector<int> min_option_widths_px = {50, 50, 50};

    auto instructions = new QuHeading(xstring("instructions"));
    auto grid = buildGrid(FIRST_Q, LAST_Q, options);
    grid->setMinimumWidthInPixels(min_width_px, min_option_widths_px);

    QVector<QuElement*> elements{
        instructions,
        grid,
    };

    QuPagePtr page((new QuPage(elements))->setTitle(xstring("title")));

    auto questionnaire = new Questionnaire(m_app, {page});
    questionnaire->setType(QuPage::PageType::Patient);
    questionnaire->setReadOnly(read_only);

    return questionnaire;
}

QuMcqGrid* Paradise24::buildGrid(
    int first_qnum, int last_qnum, const NameValueOptions options
)
{
    QVector<QuestionWithOneField> q_field_pairs;

    for (int qnum = first_qnum; qnum <= last_qnum; qnum++) {
        const QString& fieldname = Q_PREFIX + QString::number(qnum);
        const QString& description = xstring(fieldname);

        q_field_pairs.append(
            QuestionWithOneField(description, fieldRef(fieldname))
        );
    }

    auto grid = new QuMcqGrid(q_field_pairs, options);
    // Repeat options every six lines
    QVector<McqGridSubtitle> subtitles{
        {6, ""},
        {12, ""},
        {18, ""},
    };
    grid->setSubtitles(subtitles);

    const int question_width = 4;
    const QVector<int> option_widths = {1, 1, 1};
    grid->setWidth(question_width, option_widths);
    grid->setQuestionsBold(false);

    return grid;
}
